### PR TITLE
Added user configurable logging

### PR DESCRIPTION
**Pull request by Caleb-Hill**
Added a "log" element in the Config struct. Logging can be tailored by
module to log different information, at different levels, to different
places. Currently stdOut, File, and Rotating File are supported for
logging targets.

### CODE REVIEW

File: splinterd/Cargo.toml [hunk 1]
```diff
@@ -41,6 +41,7 @@ cylinder = "0.2"
 flexi_logger = "0.14"
 health = { path = "../services/health", optional = true }
 log = "0.4"
+log4rs = {version ="1.0", features = ["pattern_encoder","delete_roller","compound_policy","size_trigger","file_appender","toml_format","config_parsing"]}
```
Comment (vaporos) [inline splinterd/Cargo.toml:44]: This should be optional for now and pulled in with the experimental feature.

Note the spacing is incorrect -- "{version" vs. "{ version" and speaces after commas, etc. to match the style used elsewhere.

Version should be set to "1" not "1.0" as they are functionally equivalent and we prefer the shorter.

Are all of these features needed explicitly? This is very strange.

File: splinterd/Cargo.toml [hunk 2]
```diff
@@ -152,6 +154,9 @@ service-arg-validation = [
 ]
 service-endpoint = []
 ws-transport = ["splinter/ws-transport"]
+log-config = ["log-redirect","log-filter"]
+log-redirect = []
```
Comment (vaporos) [inline splinterd/Cargo.toml:158]: log-filter and log-redirect are not used in this PR - please remove.

File: splinterd/src/config/mod.rs [hunk 3]
```diff
@@ -799,6 +819,14 @@ impl Config {
                 debug!("Config: metrics_password: <HIDDEN> (source: {:?})", source,);
             }
         }
+        #[cfg(feature = "log-config")]
+        {
+            debug!(
+                "Config: enable_biome_credentials: {:?} (source: {:?})",
```
Comment (vaporos) [inline splinterd/src/config/mod.rs:825]: This appears to be a cut-n-paste error

File: splinterd/src/config/toml.rs [hunk 4]
```diff
@@ -16,6 +16,8 @@
 
 use crate::config::PartialConfigBuilder;
 use crate::config::{ConfigError, ConfigSource, PartialConfig};
+#[cfg(feature = "log-config")]
+use log4rs::config::RawConfig as LogConfig;
```
Comment (vaporos) [inline splinterd/src/config/toml.rs:20]: Remove use of log4rs from within crate::config and define the appropriate structs explicitly. This is important for two reasons: 1) log4rs is fundamentally unrelated to configuration loading; and 2) the config file is a versioned API and changes to it need to be handled in a backward-compatible way -- so we can not introduce an external data structure as part of our API definition.

Then, write an Into trait allowing conversion of our LogConfig struct to log4rs's RawConfig. This should maybe go in a crate::log module.

File: splinterd/src/main.rs [hunk 5]
```diff
@@ -25,8 +25,11 @@ mod error;
 mod routes;
 mod transport;
 
-use flexi_logger::{style, DeferredNow, LogSpecBuilder, Logger};
+use flexi_logger::{style, DeferredNow};
```
Comment (vaporos) [inline splinterd/src/main.rs:28]: flexi_logger should be completely disabled when log4rs is enabled.

File: splinterd/src/main.rs [hunk 6]
```diff
@@ -443,25 +446,52 @@ fn main() {
...
-    log_spec_builder.module("tokio", log::LevelFilter::Warn);
-    #[cfg(feature = "https-bind")]
-    log_spec_builder.module("h2", log::LevelFilter::Warn);
-
-    Logger::with(log_spec_builder.build())
-        .format(log_format)
-        .log_target(flexi_logger::LogTarget::StdOut)
-        .start()
-        .expect("Failed to create logger");
+    #[cfg(feature = "log-config")]
```
Comment (vaporos) [inline splinterd/src/main.rs:449]: This doesn't seem to take clap's log level into account. Similarly, hyper, tokio, h2 log-levels aren't adjusted.

File: splinterd/src/main.rs [hunk 7]
```diff
@@ -443,25 +446,52 @@ fn main() {
...
+                    Some(
+                        log4rs::init_raw_config(log_config.clone())
+                            .expect("Failed to create logger"),
+                    )
+                } else {
+                    None
+                }
+            }
+            Err(e) => {
+                println!("Could not create config: {}", e);
```
Comment (vaporos) [inline splinterd/src/main.rs:470]: Errors must be printed with eprintln! and should begin with "Error: ".

File: splinterd/src/main.rs [hunk 8]
```diff
@@ -500,13 +530,18 @@ fn setup_metrics_recorder(config: &Config) -> Result<(), UserError> {
     Ok(())
 }
 
-fn start_daemon(matches: ArgMatches) -> Result<(), UserError> {
-    // get provided config file or search default location
-    let config_file = matches
+fn get_config_file<'a>(matches: &'a ArgMatches) -> &'a str {
```
Comment (vaporos) [inline splinterd/src/main.rs:533]: This change could be a separate commit prior to the larger commit.

File: splinterd/src/main.rs (hunk 6 above)
Comment (vaporos) [inline splinterd/src/main.rs:449]: Switching to log4rs but maintaining the same functionality could be a commit prior to the larger config commit.

**Review by vaporos — CHANGES_REQUESTED** [review]

File: splinterd/Cargo.toml (hunk 1 above)
Comment (Caleb-Hill) [reply splinterd/Cargo.toml]: fixed

File: splinterd/Cargo.toml (hunk 2 above)
Comment (Caleb-Hill) [reply splinterd/Cargo.toml]: fixed

File: splinterd/src/config/mod.rs (hunk 3 above)
Comment (Caleb-Hill) [reply splinterd/src/config/mod.rs:827]: fixed

File: splinterd/src/config/toml.rs (hunk 4 above)
Comment (Caleb-Hill) [reply splinterd/src/config/toml.rs]: fixed/updated

File: splinterd/src/main.rs (hunk 5 above)
Comment (Caleb-Hill) [reply splinterd/src/main.rs]: fixed

File: splinterd/src/main.rs (hunk 6 above)
Comment (Caleb-Hill) [reply splinterd/src/main.rs:460]: fixed

File: splinterd/src/main.rs (hunk 8 above)
Comment (Caleb-Hill) [reply splinterd/src/main.rs:552]: fixed

File: splinterd/src/main.rs (hunk 6 above)
Comment (Caleb-Hill) [reply splinterd/src/main.rs:460]: fixed

Comment (Caleb-Hill) [reply splinterd/src/main.rs:460]: fixed/refactored

File: splinterd/src/main.rs (hunk 7 above)
Comment (Caleb-Hill) [reply splinterd/src/main.rs]: Changed to an expect to keep with the current behavior of stopping the app if configs are broken. Maybe it makes sense to start the app without logging but don't think its a good idea

File: splinterd/src/config/mod.rs (hunk 3 above)
Comment (peterschwarz) [reply splinterd/src/config/mod.rs:827]: This looks like it hasn't been fixed

**Review by peterschwarz — CHANGES_REQUESTED** [review]
I wonder if the log config should also have a set of partial configs, where it can generate these from cli args or a toml file (like the general configuration). That would simplify the usage in main and have the config resolution occur when collapsing the partial configs.

**Comment by Caleb-Hill** [comment]
> I wonder if the log config should also have a set of partial configs, where it can generate these from cli args or a toml file (like the general configuration). That would simplify the usage in main and have the config resolution occur when collapsing the partial configs.

While generally I think this is a good idea representing the log configs as command line args would be hugely complicated. Not undoable but really hard for anyone to use for sure. If you'd like me to refactor so most of the main.rs changes as they stand now are managed in partial.rs I can do that.

I fixed the wrong debug message thing, it won't let me comment on it for whatever reason.

File: splinterd/src/config/mod.rs (hunk 3 above)
Comment (Caleb-Hill) [reply splinterd/src/config/mod.rs]: Double super fixed

**Comment by peterschwarz** [comment]
> While generally I think this is a good idea representing the log configs as command line args would be hugely complicated. Not undoable but really hard for anyone to use for sure. If you'd like me to refactor so most of the main.rs changes as they stand now are managed in partial.rs I can do that.

The intent isn't to represent all of the possible options in every config, but to use the same mechanism, such that adding other logging overrides to the CLI args would be simple.  I'm not requesting that all the configuration options are added to the CLI args.

**Comment by Caleb-Hill** [comment]
> > While generally I think this is a good idea representing the log configs as command line args would be hugely complicated. Not undoable but really hard for anyone to use for sure. If you'd like me to refactor so most of the main.rs changes as they stand now are managed in partial.rs I can do that.
> 
> The intent isn't to represent all of the possible options in every config, but to use the same mechanism, such that adding other logging overrides to the CLI args would be simple. I'm not requesting that all the configuration options are added to the CLI args.

Yeah that makes sense, Ill update the PR once I make the change

File: splinterd/src/config/error.rs [hunk 9]
```diff
@@ -12,6 +12,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#[cfg(feature = "log-config")]
+use log4rs::config::runtime::ConfigErrors;
```
Comment (peterschwarz) [inline splinterd/src/config/error.rs:16]: Is this change still relevant?

File: splinterd/src/config/error.rs [hunk 10]
```diff
@@ -21,12 +23,17 @@ use toml::de::Error as TomlError;
 #[derive(Debug)]
 /// General error type used during `Config` contruction.
 pub enum ConfigError {
-    ReadError { file: String, err: io::Error },
+    ReadError {
```
Comment (peterschwarz) [inline splinterd/src/config/error.rs:26]: This change is not mentioned in the commit message - could be a separate commit, or was added by accident.

File: splinterd/src/main.rs [hunk 11]
```diff
@@ -460,33 +468,63 @@ fn main() {
...
-    let log_config = config.build(Root::builder().appender("stdout").build(log_level));
-    if let Ok(lc) = log_config {
-        let _handle = log4rs::init_config(lc);
-    } else {
-        unreachable!();
-        // log_config  should always return an Ok variant.
-        // Err is returned if multiple logger facades are registered which shouldn't ever happen
-    }
+            if let Ok(lc) = conf {
+                log4rs::init_config(lc).unwrap()
```
Comment (peterschwarz) [inline splinterd/src/main.rs:519]: Remove the unwrap calls, here and elsewhere

File: splinterd/src/main.rs [hunk 12]
```diff
@@ -538,7 +577,7 @@ fn get_log_filter_level(matches: &ArgMatches) -> log::LevelFilter {
     }
 }
 
-fn start_daemon(matches: ArgMatches) -> Result<(), UserError> {
+fn start_daemon(matches: ArgMatches, _log_handle: Handle) -> Result<(), UserError> {
```
Comment (peterschwarz) [inline splinterd/src/main.rs:580]: Why the unused param?

Comment (Caleb-Hill) [reply splinterd/src/main.rs:580]: Its only used with experimental turned on and throws a warning otherwise

File: splinterd/src/config/error.rs (hunk 10 above)
Comment (Caleb-Hill) [reply splinterd/src/config/error.rs:26]: Ahh yeah rustfmt  did it for me, Ill pull it out

File: splinterd/src/config/error.rs (hunk 9 above)
Comment (Caleb-Hill) [reply splinterd/src/config/error.rs:16]: Doesn't look like it is, Ill take it out

File: splinterd/src/logging/mod.rs [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+pub struct AppenderConfig {
+    pub name: String,
+    #[serde(default = "default_pattern")]
+    pub encoder: String,
+    pub target: LogTarget,
+}
+
+#[derive(Deserialize, Clone, Debug)]
+pub enum LogTarget {
+    StdOut,
```
[the remark this answers is not in the document]
Comment (Caleb-Hill) [reply splinterd/src/logging/mod.rs:36]: added some aliases that seemed reasonable-ish

File: splinterd/src/main.rs (hunk 11 above)
Comment (Caleb-Hill) [reply splinterd/src/main.rs]: No more just raw unwrap calls

File: splinterd/src/config/mod.rs [hunk 14]
```diff
@@ -553,6 +563,37 @@ impl Config {
             None
         }
     }
+    #[cfg(feature = "log-config")]
+    pub fn root_logger(&self) -> &RootConfig {
+        &self.root_logger.0
+    }
+    #[cfg(feature = "log-config")]
```
Comment (vaporos) [inline splinterd/src/config/mod.rs:570]: Missing a empty line above this line.

File: splinterd/src/config/partial.rs [hunk 15]
```diff
@@ -785,4 +821,45 @@ impl PartialConfig {
...
+    /// Adds a `verbosity` value to the `PartialConfig` object.
+    ///
+    /// # Arguments
+    ///
+    ///  * `level` - The log level the system will use for the root_logger logger
+    pub fn with_verbosity(mut self, level: Option<log::Level>) -> Self {
+        self.verbosity = level;
+        self
+    }
+    #[cfg(feature = "log-config")]
```
Comment (vaporos) [inline splinterd/src/config/partial.rs:835]: Missing a newline above this line (also, same with functions below in this chunk).

File: splinterd/src/logging/mod.rs [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+    StdErr,
```
Comment (vaporos) [inline splinterd/src/logging/mod.rs:37]: s/StdErr/stderr/ and s/StdOut/Stdout/

File: splinterd/src/logging/mod.rs [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+#[derive(Deserialize, Clone, Debug)]
+pub struct LogConfigPartial {
+    root: Option<RootConfig>,
+    appenders: Option<Vec<AppenderConfig>>,
+    loggers: Option<Vec<LoggerConfig>>,
+}
+
+pub fn default_pattern() -> String {
```
Comment (vaporos) [inline splinterd/src/logging/mod.rs:49]: This should at most be a const declaration.

File: splinterd/src/logging/mod.rs [hunk 18]
```diff
@@ -28,13 +36,20 @@ pub struct LogConfig {
 pub struct LoggerConfig {
     name: String,
     appenders: Vec<String>,
-    filter: Level,
+    level: Level,
 }
 
-#[derive(Clone, Debug)]
+#[derive(Deserialize, Clone, Debug)]
+pub struct UnNamedLoggerConfig {
```
Comment (peterschwarz) [inline splinterd/src/logging/mod.rs:43]: It's unnamed, not un named, so remove the capital N

File: splinterd/src/config/toml.rs [hunk 19]
```diff
@@ -195,6 +203,21 @@ impl PartialConfigBuilder for TomlPartialConfigBuilder {
                 .with_metrics_username(self.toml_config.metrics_username)
                 .with_metrics_password(self.toml_config.metrics_password)
         }
+        #[cfg(feature = "log-config")]
+        {
+            if let Some(mut loggers) = self.toml_config.loggers {
+                if let Some(un_named) = loggers.remove("root") {
```
Comment (peterschwarz) [inline splinterd/src/config/toml.rs:209]: unnamed, not un_named

File: splinterd/src/main.rs [hunk 20]
```diff
@@ -462,33 +469,65 @@ fn main() {
...
+                    encoder: String::from(DEFAULT_PATTERN),
+                    kind: LogTarget::Stdout,
+                }],
+                loggers: vec![],
+            };
+            log4rs::init_config(
+                //Always produces a valid config try_into_config just returns a Result
+                default_config
+                    .try_into()
+                    .expect("Default logging config failure"),
```
Comment (peterschwarz) [inline splinterd/src/main.rs:492]: It might be more clear to use an `unreachable!()` statement like below

File: splinterd/src/logging/mod.rs [hunk 21]
```diff
@@ -28,13 +36,20 @@ pub struct LogConfig {
...
+    appenders: Vec<String>,
+    #[serde(alias = "filter")]
```
Comment (peterschwarz) [inline splinterd/src/logging/mod.rs:45]: Should we alias this?  What if we want to use "filter" for ignoring modules?

**Review by peterschwarz — APPROVED** [review]

File: splinterd/src/logging/mod.rs (hunk 17 above)
Comment (Caleb-Hill) [reply splinterd/src/logging/mod.rs]: It needs to be a function for its use within serde, so I have made the function private and added a public const with the actual pattern for use outside of this file.

File: splinterd/src/logging/mod.rs (hunk 16 above)
Comment (Caleb-Hill) [reply splinterd/src/logging/mod.rs]: fixed

File: splinterd/src/config/partial.rs (hunk 15 above)
Comment (Caleb-Hill) [reply splinterd/src/config/partial.rs:838]: fixed all the spacing issues

**Review by vaporos — APPROVED** [review]
